core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java

```diff
@@ -1491,6 +1496,43 @@ private static class PeriodOperandTypeChecker
     }
   }
 
+  /**
+   * Parameter type-checking strategy where types must be Map
+   * and adjust their types to suit the Map function's constructor.
+   */
+  private static class MapSparkFunctionOperandTypeChecker extends SameOperandTypeChecker {
+    MapSparkFunctionOperandTypeChecker() {
+      super(1);
+    }
+
+    @Override public boolean checkOperandTypes(
+        SqlCallBinding callBinding,
+        boolean throwOnFailure) {
+
+      final SqlNode op0 = callBinding.operand(0);
+      if (!OperandTypes.MAP.checkSingleOperandType(
+          callBinding,
+          op0,
+          0,
+          throwOnFailure)) {
+        return false;
+      }
+
+      final RelDataType mapKeyType =
+          getKeyTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op0));
+      final RelDataType mapValueType =
+          getValueTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op0));
+
+      // If the key and value types are not unknown,
+      // adjust their types to suit the Map function's constructor.
+      if (mapKeyType.getSqlTypeName() != SqlTypeName.UNKNOWN
```

Comment:
I think we have discussed this in the past: I don't think that the type checker is supposed to modify the node that is being checked.

There are type checking strategies that will backtrack (e.g., SqlOperandTypeChecker.or), and they do not expect the checked data structure to change.

I think this pattern is very dangerous, and can lead to hard to diagnose bugs.

On the other hand, I don't know how this should be done properly. The map function does need to introduce type casts. But maybe the typechecking stage is not the right stage.

Reply:
Currently, this type check is specially prepared for Spark's map function, but there may be some problems that are difficult to eliminate as you said. However, Spark currently performs some type conversions during parameter checking. I have to admit that Spark does not do this very well. 
 
I believe that after I have a deeper understanding of Calcite and the type system in the future, I can improve this part. I hope that day will not be too far away.

Reply:
As Julian said in the Jira comments, you should insert casts during validation, before type checking.

Reply:
ok, thank you.